File: sdks/python/costtrack.py
"""
CostTrack Python SDK - Auto-instrument LLM calls.

Usage:
    import costtrack  # Add this one line at the top of your app

    # All Anthropic/OpenAI calls are now tracked automatically.
    # Configure via environment variables:
    #   COSTTRACK_PROXY_URL=http://localhost:8080
    #   COSTTRACK_TEAM=my-team
    #   COSTTRACK_APP=my-app
    #   COSTTRACK_FEATURE=my-feature
    #   COSTTRACK_CUSTOMER_TIER=enterprise
    #   COSTTRACK_SESSION_ID=session-123  (optional)
"""

import contextlib
import logging
import os
import threading

logger = logging.getLogger("costtrack")

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

_config = {
    "proxy_url": os.environ.get("COSTTRACK_PROXY_URL", "http://localhost:8080"),
    "team": os.environ.get("COSTTRACK_TEAM", ""),
    "app": os.environ.get("COSTTRACK_APP", ""),
    "feature": os.environ.get("COSTTRACK_FEATURE", ""),
    "customer_tier": os.environ.get("COSTTRACK_CUSTOMER_TIER", ""),
    "session_id": os.environ.get("COSTTRACK_SESSION_ID", ""),
}

# Thread-local storage for session context.
_local = threading.local()


def configure(
    proxy_url: str = "",
    team: str = "",
    app: str = "",
    feature: str = "",
    customer_tier: str = "",
    session_id: str = "",
) -> None:
    """Explicitly configure CostTrack. Overrides environment variables."""
    if proxy_url:
        _config["proxy_url"] = proxy_url
    if team:
        _config["team"] = team
    if app:
        _config["app"] = app
    if feature:
        _config["feature"] = feature
    if customer_tier:
        _config["customer_tier"] = customer_tier
    if session_id:
        _config["session_id"] = session_id

    # Re-patch with updated config so new clients pick up the changes.
    _patch_all()


@contextlib.contextmanager
def session(session_id: str):
    """Context manager that tags all LLM calls within the block with a session ID.

    Usage:
        with costtrack.session("user-query-123"):
            response = client.messages.create(...)
    """
    previous = getattr(_local, "session_id", None)
    _local.session_id = session_id
    try:
        yield
    finally:
        _local.session_id = previous


# ---------------------------------------------------------------------------
# Header building
# ---------------------------------------------------------------------------

def _build_headers() -> dict:
    """Build the X-CostTrack-* headers from current config and thread-local state."""
    headers = {}
    mapping = {
        "X-CostTrack-Team": "team",
        "X-CostTrack-App": "app",
        "X-CostTrack-Feature": "feature",
        "X-CostTrack-Customer-Tier": "customer_tier",
        "X-CostTrack-Session-ID": "session_id",
    }
    for header, key in mapping.items():
        value = _config.get(key, "")
        if value:
            headers[header] = value

    # Thread-local session overrides the global config.
    thread_session = getattr(_local, "session_id", None)
    if thread_session:
        headers["X-CostTrack-Session-ID"] = thread_session

    return headers


# ---------------------------------------------------------------------------
# Anthropic patching
# ---------------------------------------------------------------------------

_anthropic_original_init = None


def _patch_anthropic():
    """Monkey-patch the Anthropic SDK to route through the CostTrack proxy."""
    global _anthropic_original_init

    try:
        import anthropic
    except ImportError:
        logger.debug("anthropic package not installed, skipping patch")
        return

    if _anthropic_original_init is None:
        _anthropic_original_init = anthropic.Anthropic.__init__

    original_init = _anthropic_original_init

    def patched_init(self, *args, **kwargs):
        # Set base_url to proxy unless the caller explicitly provided one.
        if "base_url" not in kwargs:
            proxy = _config["proxy_url"].rstrip("/")
            kwargs["base_url"] = proxy + "/"

        # Merge CostTrack headers into default_headers.
        costtrack_headers = _build_headers()
        if costtrack_headers:
            existing = kwargs.get("default_headers") or {}
            merged = {**costtrack_headers, **existing}
            kwargs["default_headers"] = merged

        original_init(self, *args, **kwargs)

    anthropic.Anthropic.__init__ = patched_init
    logger.info("CostTrack: patched anthropic.Anthropic")


# ---------------------------------------------------------------------------
# OpenAI patching
# ---------------------------------------------------------------------------

_openai_original_init = None


def _patch_openai():
    """Monkey-patch the OpenAI SDK to route through the CostTrack proxy."""
    global _openai_original_init

    try:
        import openai
    except ImportError:
        logger.debug("openai package not installed, skipping patch")
        return

    if _openai_original_init is None:
        _openai_original_init = openai.OpenAI.__init__

    original_init = _openai_original_init

    def patched_init(self, *args, **kwargs):
        # Set base_url to proxy unless the caller explicitly provided one.
        if "base_url" not in kwargs:
            proxy = _config["proxy_url"].rstrip("/")
            kwargs["base_url"] = proxy + "/"

        # Merge CostTrack headers into default_headers.
        costtrack_headers = _build_headers()
        if costtrack_headers:
            existing = kwargs.get("default_headers") or {}
            merged = {**costtrack_headers, **existing}
            kwargs["default_headers"] = merged

        original_init(self, *args, **kwargs)

    openai.OpenAI.__init__ = patched_init
    logger.info("CostTrack: patched openai.OpenAI")


# ---------------------------------------------------------------------------
# Auto-patch on import
# ---------------------------------------------------------------------------

def _patch_all():
    """Attempt to patch all supported LLM SDKs."""
    for patcher in (_patch_anthropic, _patch_openai):
        try:
            patcher()
        except Exception as exc:
            logger.warning("CostTrack: failed to patch: %s", exc)


# Run automatically on import.
_patch_all()

File: sdks/node/costtrack.ts
/**
 * CostTrack Node SDK - Auto-instrument LLM calls.
 *
 * Usage:
 *   import 'costtrack'  // Add this one line
 *
 * Or with configuration:
 *   import { configure, withSession } from './costtrack'
 *   configure({ team: 'my-team', app: 'my-app' })
 *
 * Configure via environment variables:
 *   COSTTRACK_PROXY_URL=http://localhost:8080
 *   COSTTRACK_TEAM=my-team
 *   COSTTRACK_APP=my-app
 *   COSTTRACK_FEATURE=my-feature
 *   COSTTRACK_CUSTOMER_TIER=enterprise
 *   COSTTRACK_SESSION_ID=session-123  (optional)
 *   COSTTRACK_TRACE_ID=trace-456  (optional)
 *   COSTTRACK_PROMPT_TEMPLATE=template-id  (optional)
 */

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

interface CostTrackConfig {
  proxyUrl: string;
  team: string;
  app: string;
  feature: string;
  customerTier: string;
  sessionId: string;
  traceId: string;
  promptTemplate: string;
}

const config: CostTrackConfig = {
  proxyUrl: process.env.COSTTRACK_PROXY_URL || "http://localhost:8080",
  team: process.env.COSTTRACK_TEAM || "",
  app: process.env.COSTTRACK_APP || "",
  feature: process.env.COSTTRACK_FEATURE || "",
  customerTier: process.env.COSTTRACK_CUSTOMER_TIER || "",
  sessionId: process.env.COSTTRACK_SESSION_ID || "",
  traceId: process.env.COSTTRACK_TRACE_ID || "",
  promptTemplate: process.env.COSTTRACK_PROMPT_TEMPLATE || "",
};

// Session ID and trace ID overrides for withSession().
let _activeSessionId: string | null = null;
let _activeTraceId: string | null = null;

export function configure(opts: Partial<Omit<CostTrackConfig, "proxyUrl"> & { proxyUrl: string }>): void {
  if (opts.proxyUrl) config.proxyUrl = opts.proxyUrl;
  if (opts.team) config.team = opts.team;
  if (opts.app) config.app = opts.app;
  if (opts.feature) config.feature = opts.feature;
  if (opts.customerTier) config.customerTier = opts.customerTier;
  if (opts.sessionId) config.sessionId = opts.sessionId;
  if (opts.traceId) config.traceId = opts.traceId;
  if (opts.promptTemplate) config.promptTemplate = opts.promptTemplate;

  // Re-patch so new clients pick up changes.
  patchAll();
}

/**
 * Run a function with a specific session ID attached to all LLM calls.
 *
 * Usage:
 *   await withSession("user-query-123", async () => {
 *     const response = await client.messages.create({ ... });
 *   });
 *
 *   await withSession("session-123", async () => { ... }, { traceId: "trace-456" });
 */
export async function withSession<T>(
  sessionId: string,
  fn: () => T | Promise<T>,
  opts?: { traceId?: string },
): Promise<T> {
  const previousSession = _activeSessionId;
  const previousTrace = _activeTraceId;
  _activeSessionId = sessionId;
  if (opts?.traceId) {
    _activeTraceId = opts.traceId;
  }
  try {
    return await fn();
  } finally {
    _activeSessionId = previousSession;
    _activeTraceId = previousTrace;
  }
}

// ---------------------------------------------------------------------------
// Header building
// ---------------------------------------------------------------------------

function buildHeaders(): Record<string, string> {
  const headers: Record<string, string> = {};

  if (config.team) headers["X-CostTrack-Team"] = config.team;
  if (config.app) headers["X-CostTrack-App"] = config.app;
  if (config.feature) headers["X-CostTrack-Feature"] = config.feature;
  if (config.customerTier) headers["X-CostTrack-Customer-Tier"] = config.customerTier;
  if (config.promptTemplate) headers["X-CostTrack-Prompt-Template"] = config.promptTemplate;

  // Active session from withSession() overrides config.
  const sessionId = _activeSessionId || config.sessionId;
  if (sessionId) headers["X-CostTrack-Session-ID"] = sessionId;

  // Active trace ID from withSession() overrides config.
  const traceId = _activeTraceId || config.traceId;
  if (traceId) headers["X-CostTrack-Trace-ID"] = traceId;

  return headers;
}

// ---------------------------------------------------------------------------
// Anthropic patching
// ---------------------------------------------------------------------------

let _anthropicOriginalConstructor: (new (...args: any[]) => any) | null = null;

function patchAnthropic(): void {
  let anthropicModule: any;
  try {
    anthropicModule = require("@anthropic-ai/sdk");
  } catch {
    // Package not installed.
    return;
  }

  const OriginalAnthropic = _anthropicOriginalConstructor || anthropicModule.default?.Anthropic || anthropicModule.Anthropic;
  if (!OriginalAnthropic) return;

  if (!_anthropicOriginalConstructor) {
    _anthropicOriginalConstructor = OriginalAnthropic;
  }

  function PatchedAnthropic(this: any, opts: any = {}) {
    // Set baseURL to proxy unless the caller explicitly provided one.
    if (!opts.baseURL) {
      opts.baseURL = config.proxyUrl.replace(/\/+$/, "") + "/";
    }

    // Merge CostTrack headers into defaultHeaders.
    const costtrackHeaders = buildHeaders();
    opts.defaultHeaders = { ...costtrackHeaders, ...(opts.defaultHeaders || {}) };

    return new (_anthropicOriginalConstructor as any)(opts);
  }

  // Copy static properties and prototype.
  Object.setPrototypeOf(PatchedAnthropic, OriginalAnthropic);
  PatchedAnthropic.prototype = OriginalAnthropic.prototype;

  // Replace on the module.
  if (anthropicModule.default?.Anthropic) {
    anthropicModule.default.Anthropic = PatchedAnthropic;
  }
  if (anthropicModule.Anthropic) {
    anthropicModule.Anthropic = PatchedAnthropic;
  }

  console.log("CostTrack: patched @anthropic-ai/sdk");
}

// ---------------------------------------------------------------------------
// OpenAI patching
// ---------------------------------------------------------------------------

let _openaiOriginalConstructor: (new (...args: any[]) => any) | null = null;

function patchOpenAI(): void {
  let openaiModule: any;
  try {
    openaiModule = require("openai");
  } catch {
    // Package not installed.
    return;
  }

  const OriginalOpenAI = _openaiOriginalConstructor || openaiModule.default?.OpenAI || openaiModule.OpenAI;
  if (!OriginalOpenAI) return;

  if (!_openaiOriginalConstructor) {
    _openaiOriginalConstructor = OriginalOpenAI;
  }

  function PatchedOpenAI(this: any, opts: any = {}) {
    // Set baseURL to proxy unless the caller explicitly provided one.
    if (!opts.baseURL) {
      opts.baseURL = config.proxyUrl.replace(/\/+$/, "") + "/";
    }

    // Merge CostTrack headers into defaultHeaders.
    const costtrackHeaders = buildHeaders();
    opts.defaultHeaders = { ...costtrackHeaders, ...(opts.defaultHeaders || {}) };

    return new (_openaiOriginalConstructor as any)(opts);
  }

  // Copy static properties and prototype.
  Object.setPrototypeOf(PatchedOpenAI, OriginalOpenAI);
  PatchedOpenAI.prototype = OriginalOpenAI.prototype;

  // Replace on the module.
  if (openaiModule.default?.OpenAI) {
    openaiModule.default.OpenAI = PatchedOpenAI;
  }
  if (openaiModule.OpenAI) {
    openaiModule.OpenAI = PatchedOpenAI;
  }

  console.log("CostTrack: patched openai");
}

// ---------------------------------------------------------------------------
// Auto-patch on import
// ---------------------------------------------------------------------------

function patchAll(): void {
  for (const patcher of [patchAnthropic, patchOpenAI]) {
    try {
      patcher();
    } catch (err) {
      console.warn("CostTrack: failed to patch:", err);
    }
  }
}

// Run automatically on import.
patchAll();
